Add test for 256-color configuration values

yurikhan opened the pull request:
##### SUMMARY

See #78607.

##### ISSUE TYPE

- Test Pull Request

##### COMPONENT NAME

config

yurikhan committed:
Add test for 256-color configuration values

See #78607.

webknjaz commented on test/units/config/test_manager.py:
```diff
@@ -142,3 +143,16 @@ def decrypt(self, value, filename=None, obj=None):
         assert actual_value == "vault text"
+
+
+@pytest.mark.parametrize(("key", "expected_value"), (
+    ("COLOR_UNREACHABLE", "bright red"),
+    ("COLOR_VERBOSE", "rgb013"),
+    ("COLOR_DEBUG", "gray10")))
```
If this is desired to be merged as a regression test before fixing, maybe use the xfail marker?
```suggestion
@pytest.mark.parametrize(
    ("key", "expected_value"),
    (
        ("COLOR_UNREACHABLE", "bright red"),
        pytest.param(
            "COLOR_VERBOSE", "rgb013",
            marks=pytest.mark.xfail(reason="https://github.com/ansible/ansible/issues/78607", strict=True),
        ),
        pytest.param(
            "COLOR_DEBUG", "gray10",
            marks=pytest.mark.xfail(reason="https://github.com/ansible/ansible/issues/78607", strict=True),
        ),
    ),
)
```

bcoca committed:
color is not restricted to 16 choices

 currently supports up to 256, not listing them all
 TOOD: create examples and point to/list the basic 16

bcoca commented:
hmm, shouldn't the test be failing?

sivel commented:
This commit https://github.com/ansible/ansible/pull/78613/commits/03654a2a499b34c385efafeb3379db4c16101c38 effectively disabled the tests.

yurikhan commented:
Yeah, the `xfail` mark means “expected to fail”. The fixer of the bug is supposed to remove the mark.

yurikhan commented:
I assume by “my PR” you’re referring to #78617, but you accidentally the verb. Do you mean “add the commit from #78617 and revert the xfail”?

webknjaz commented:
> I assume by “my PR” you’re referring to #78617, but you accidentally the verb. Do you mean “add the commit from #78617 and revert the xfail”?

I suppose so. You could first cherry-pick Brian's commit, ensure that the test starts failing (because xfail_strict produces a failure on XPASS — https://blog.ganssle.io/articles/2021/11/pytest-xfail.html). And then, do `git revert` as a finishing bit.

webknjaz commented:
> This commit [03654a2](https://github.com/ansible/ansible/commit/03654a2a499b34c385efafeb3379db4c16101c38) effectively disabled the tests.

It did not. xfail is not the same as skip. It does run the test (unless you set `run=False`) and then it makes sure that it fails (`strict=True`) makes it fail when it starts passing so when the problem is actually fixed, you get notified loudly.

bcoca commented:
@yurikhan yes, just add my commit/copy the code here and revert the xfail change

webknjaz commented:
/azp run

mattclay commented:
@yurikhan Can you add a changelog fragment to document the change?

yurikhan commented:
Do test suite changes deserve a changelog entry, or should I describe the bug fixed instead?

Will look into it this coming weekend.

mattclay commented:
@yurikhan Only the bug fix requires a changelog entry. Unit test changes should not be in the changelog.

bcoca commented:
/azp run

bcoca committed:
added clog

mattclay commented on changelogs/fragments/colors.yml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - ansilbe terminal color settings were incorrectly limited to 16 options via 'choices', removing so all 256 can be accessed.
```
```suggestion
  - ansible terminal color settings were incorrectly limited to 16 options via 'choices', removing so all 256 can be accessed.
```

mattclay committed:
Fix changelog fragment typo
